### PR TITLE
PLIC

### PR DESCRIPTION
**Pull request by laanwj**
Based on `lib/drivers/include/plic.h` in SDK and loosely on e310x svd.

### CODE REVIEW

File: k210.svd [hunk 1]
```diff
@@ -39,12 +39,106 @@
...
+                    <dimIncrement>0x04</dimIncrement>
+                    <name>pending[%s]</name>
+                    <description>Interrupt Pending Register</description>
+                    <addressOffset>0x001000</addressOffset>
                 </register>
+                <cluster>
+                    <name>target_enables[%s]</name>
+                    <description>Target Interrupt Enables</description>
+                    <addressOffset>0x002000</addressOffset>
+                    <dim>2</dim>
```
Comment (Disasm) [inline k210.svd]: Are there only two targets in K210?

**Review by Disasm — COMMENTED** [review]
This number of targets needs to be checked, everything else LGTM

File: k210.svd (hunk 1 above)
Comment (laanwj) [reply k210.svd]: Only if the number of targets is the number of harts, which I've assumed, but I might misunderstand this part of the spec.

Comment (Disasm) [reply k210.svd]: [plic.h:288](https://github.com/kendryte/kendryte-standalone-sdk/blob/74bb90431969c55d83ce4fa2e6cd087a152b6153/lib/drivers/include/plic.h#L288)
```c
struct
{
    uint32_t enable[32 * 2];/* Offset 0x00-0x7C: Bit 0 is zero, Bits 1-1023 is bits*/
} target[15872 / 2];
```

[plic.c:35](https://github.com/kendryte/kendryte-standalone-sdk/blob/74bb90431969c55d83ce4fa2e6cd087a152b6153/lib/drivers/plic.c#L35)
```c
plic->target_enables.target[core_id].enable[i] = 0;
```
As far as I can understand this code, for `core_id=0` we use offset 0 and for `core_id=1` we use offset 256 (32*2*4), that is twice the size of enable array (1024 bits == 128 bytes (0x80)). From this I can deduce that `core_id=1` is target 3, not target 1.

On the other hand, it sounds reasonable to have two targets for each hart if you have Supervisor mode: one target for `M` mode and one for `S` mode. So my guess is that we have 4 targets: `hart 0 M mode`, `hart 0 S mode`, `hart 1 M mode`, `hart 1 S mode`. But who knows.

**Comment by laanwj** [comment]
Fairly sure you're right. I didn't grasp the `*2` and `/2` factors as first, but now it makes sense to me. It's the same for `targets`, there's `0x1000` per target, but their actual per-target structure is sized `0x2000`.

I've updated this with the assumption that there's four interrupt targets. I'll not name them but there's a good assumption they match the SiFive U54 w/ M-Mode and S-Mode per hart.

**Comment by Disasm** [comment]
Thanks!

**Pull request merged**